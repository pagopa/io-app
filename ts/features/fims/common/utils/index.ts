import { ServiceId } from "../../../../../definitions/backend/ServiceId";
import { ServicePublic } from "../../../../../definitions/backend/ServicePublic";
import { GlobalState } from "../../../../store/reducers/types";
import { serviceByIdSelector } from "../../../services/details/store/reducers";
import { fimsCtaTextSelector } from "../../singleSignOn/store/selectors";
import { trackDataShare, trackDataShareAccepted } from "../analytics";

export const computeAndTrackDataShare = (
  service: ServicePublic,
  state: GlobalState
) => {
  const ctaText = fimsCtaTextSelector(state);
  trackDataShare(
    service.service_id,
    service.service_name,
    service.organization_name,
    service.organization_fiscal_code,
    ctaText
  );
};

export const computeAndTrackDataShareAccepted = (
  serviceId: ServiceId,
  state: GlobalState
) => {
  const service = serviceByIdSelector(state, serviceId);
  const ctaText = fimsCtaTextSelector(state);
  trackDataShareAccepted(
    serviceId,
    service?.service_name,
    service?.organization_name,
    service?.organization_fiscal_code,
    ctaText
  );
};
