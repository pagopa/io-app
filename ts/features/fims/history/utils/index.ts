import { IOToast } from "@pagopa/io-app-design-system";
import { constNull } from "fp-ts/lib/function";
import { Alert } from "react-native";
import I18n from "../../../../i18n";

export const showFimsExportError = () =>
  IOToast.error(I18n.t("FIMS.history.exportData.errorToast"));

export const showFimsExportSuccess = () =>
  IOToast.success(I18n.t("FIMS.history.exportData.successToast"));

export const showFimsAlreadyExportingAlert = () =>
  Alert.alert(
    I18n.t("FIMS.history.exportData.alerts.alreadyExporting.title"),
    I18n.t("FIMS.history.exportData.alerts.alreadyExporting.body"),
    [{ text: I18n.t("global.buttons.ok"), onPress: constNull }]
  );
