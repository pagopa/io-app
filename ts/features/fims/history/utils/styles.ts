import { StyleSheet } from "react-native";

export const FimsHistorySharedStyles = StyleSheet.create({
  fixedHeightListItem: {
    justifyContent: "center",
    height: 117
  }
});
