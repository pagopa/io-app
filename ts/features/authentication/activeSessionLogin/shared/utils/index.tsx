import I18n from "i18next";
import { Millisecond } from "@pagopa/ts-commons/lib/units";
import { Platform } from "react-native";
import { CieAuthenticationErrorReason } from "../../../login/cie/store/actions/cie";
import { ReadingState } from "../../../login/cie/components/CieCardReadingAnimation";

// the timeout we sleep until move to consent form screen when authentication goes well
export const WAIT_TIMEOUT_NAVIGATION = 1700 as Millisecond;
export const WAIT_TIMEOUT_NAVIGATION_ACCESSIBILITY = 5000 as Millisecond;
export const VIBRATION = 100 as Millisecond;
export const accessibityTimeout = 100 as Millisecond;

// A subset of Cie Events (errors) which is of interest to analytics
export const analyticActions = new Map<CieAuthenticationErrorReason, string>([
  // Reading interrupted before the sdk complete the reading
  ["Transmission Error", I18n.t("authentication.cie.card.error.onTagLost")],
  ["ON_TAG_LOST", I18n.t("authentication.cie.card.error.onTagLost")],
  [
    "TAG_ERROR_NFC_NOT_SUPPORTED",
    I18n.t("authentication.cie.card.error.unknownCardContent")
  ],
  [
    "ON_TAG_DISCOVERED_NOT_CIE",
    I18n.t("authentication.cie.card.error.unknownCardContent")
  ],
  ["PIN Locked", I18n.t("authentication.cie.card.error.generic")],
  ["ON_CARD_PIN_LOCKED", I18n.t("authentication.cie.card.error.generic")],
  ["ON_PIN_ERROR", I18n.t("authentication.cie.card.error.tryAgain")],
  ["PIN_INPUT_ERROR", ""],
  ["CERTIFICATE_EXPIRED", I18n.t("authentication.cie.card.error.generic")],
  ["CERTIFICATE_REVOKED", I18n.t("authentication.cie.card.error.generic")],
  ["AUTHENTICATION_ERROR", I18n.t("authentication.cie.card.error.generic")],
  [
    "EXTENDED_APDU_NOT_SUPPORTED",
    I18n.t("authentication.cie.nfc.apduNotSupported")
  ],
  [
    "ON_NO_INTERNET_CONNECTION",
    I18n.t("authentication.cie.card.error.tryAgain")
  ],
  ["STOP_NFC_ERROR", ""],
  ["START_NFC_ERROR", ""]
]);

export type TextForState = {
  title: string;
  subtitle?: string;
  content: string;
};

// some texts changes depending on current running Platform
export const getTextForState = (
  state: ReadingState.waiting_card | ReadingState.error,
  errorMessage: string = ""
): TextForState => {
  const texts: Record<
    ReadingState.waiting_card | ReadingState.error,
    TextForState
  > = Platform.select({
    ios: {
      [ReadingState.waiting_card]: {
        title: I18n.t("authentication.cie.card.titleiOS"),
        subtitle: I18n.t("authentication.cie.card.layCardMessageHeaderiOS"),
        // the native alert hides the screen content and shows a message it self
        content: ""
      },
      [ReadingState.error]: {
        title: I18n.t("authentication.cie.card.error.readerCardLostTitle"),
        subtitle: "",
        // the native alert hides the screen content and shows a message it self
        content: ""
      },
      [ReadingState.reading]: {
        title: I18n.t("authentication.cie.card.titleiOS"),
        subtitle: I18n.t("authentication.cie.card.layCardMessageHeaderiOS"),
        // the native alert hides the screen content and shows a message it self
        content: ""
      }
    },
    default: {
      [ReadingState.waiting_card]: {
        title: I18n.t("authentication.cie.card.title"),
        subtitle: I18n.t("authentication.cie.card.layCardMessageHeader"),
        content: I18n.t("authentication.cie.card.layCardMessageFooter")
      },
      [ReadingState.error]: {
        title: I18n.t("authentication.cie.card.error.readerCardLostTitle"),
        subtitle: I18n.t("authentication.cie.card.error.onTagLost"),
        content: errorMessage
      }
    }
  });
  return texts[state];
};
