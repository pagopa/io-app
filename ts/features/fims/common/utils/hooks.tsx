import * as React from "react";
import { ServiceId } from "../../../../../definitions/backend/ServiceId";
import { useIODispatch, useIOSelector } from "../../../../store/hooks";
import { serviceByIdPotSelector } from "../../../services/details/store/reducers";
import { loadServiceDetail } from "../../../services/details/store/actions/details";
import { isStrictNone } from "../../../../utils/pot";

export const useAutoFetchingServiceByIdPot = (serviceId: ServiceId) => {
  const dispatch = useIODispatch();
  const serviceData = useIOSelector(state =>
    serviceByIdPotSelector(state, serviceId)
  );
  const shouldFetchServiceData = isStrictNone(serviceData);

  React.useEffect(() => {
    if (shouldFetchServiceData) {
      dispatch(loadServiceDetail.request(serviceId));
    }
  }, [dispatch, serviceId, shouldFetchServiceData]);

  return { serviceData };
};
