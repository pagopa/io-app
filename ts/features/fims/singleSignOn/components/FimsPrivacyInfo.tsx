import { Label } from "@pagopa/io-app-design-system";
import * as React from "react";
import { LoadingSkeleton } from "../../../../components/ui/Markdown/LoadingSkeleton";
import I18n from "../../../../i18n";
import { openWebUrl } from "../../../../utils/url";

export const FimsPrivacyInfo = ({ privacyUrl }: PrivacyInfoProps) =>
  privacyUrl !== undefined ? (
    <Label weight="Regular">
      {I18n.t("FIMS.consentsScreen.privacy1")}
      <Label asLink onPress={() => openWebUrl(privacyUrl)}>
        {I18n.t("FIMS.consentsScreen.privacyCta")}
      </Label>
    </Label>
  ) : (
    <LoadingSkeleton lines={2} />
  );

type PrivacyInfoProps = { privacyUrl?: string };
