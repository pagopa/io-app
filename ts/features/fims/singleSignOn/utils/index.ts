import * as pot from "@pagopa/ts-commons/lib/pot";
import { ActionType } from "typesafe-actions";
import { FimsFlowStateTags, FimsSSOState } from "../store/reducers";
import { startApplicationInitialization } from "../../../../store/actions/application";
import { isStrictSome } from "../../../../utils/pot";

export const foldFimsFlowState = <A>(
  flowState: FimsFlowStateTags,
  onConsents: (state: "consents") => A,
  onInAppBrowser: (state: "in-app-browser-loading") => A,
  onAbort: (state: "abort") => A,
  onShouldRestart: (state: "fastLogin_forced_restart") => A,
  onIdle: (state: "idle") => A
) => {
  switch (flowState) {
    case "abort":
      return onAbort(flowState);
    case "in-app-browser-loading":
      return onInAppBrowser(flowState);
    case "fastLogin_forced_restart":
      return onShouldRestart(flowState);
    case "idle":
      return onIdle(flowState);
  }
  return onConsents(flowState);
};

export const foldFimsFlowStateK =
  <A>(
    onConsents: (state: "consents") => A,
    onInAppBrowser: (state: "in-app-browser-loading") => A,
    onAbort: (state: "abort") => A,
    onShouldRestart: (state: "fastLogin_forced_restart") => A,
    onIdle: (state: "idle") => A
  ) =>
  (flowState: FimsFlowStateTags) =>
    foldFimsFlowState(
      flowState,
      onConsents,
      onInAppBrowser,
      onAbort,
      onShouldRestart,
      onIdle
    );

export const shouldRestartFimsAuthAfterFastLoginFailure = (
  state: FimsSSOState,
  action: ActionType<typeof startApplicationInitialization>
) => {
  const fastLoginSessionExpired = !!(
    action.payload && action.payload.handleSessionExpiration
  );
  if (fastLoginSessionExpired) {
    const hasExpiredDuringConsentsRetrieval = pot.isLoading(state.consentsData);
    const hasExpiredWhileRetrievingServiceData =
      state.currentFlowState === "consents" && isStrictSome(state.consentsData);
    const hasExpiredDuringInAppBrowserRedirectUriRetrieval =
      state.currentFlowState === "in-app-browser-loading";
    return (
      hasExpiredDuringConsentsRetrieval ||
      hasExpiredWhileRetrievingServiceData ||
      hasExpiredDuringInAppBrowserRedirectUriRetrieval
    );
  }
  return false;
};
