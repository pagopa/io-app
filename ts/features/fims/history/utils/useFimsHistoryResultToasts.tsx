import { IOToast } from "@pagopa/io-app-design-system";
import { constVoid } from "fp-ts/lib/function";
import * as React from "react";
import { Alert } from "react-native";
import * as RemoteValue from "../../../../common/model/RemoteValue";
import I18n from "../../../../i18n";
import { useIODispatch, useIOSelector } from "../../../../store/hooks";
import {
  fimsHistoryExport,
  resetFimsHistoryExportState
} from "../store/actions";
import { fimsHistoryExportStateSelector } from "../store/selectors";

const showFimsExportError = () =>
  IOToast.error(I18n.t("FIMS.history.exportData.errorToast"));

const showFimsExportSuccess = () =>
  IOToast.success(I18n.t("FIMS.history.exportData.successToast"));

const showFimsAlreadyExportingAlert = (onPress: () => void) =>
  Alert.alert(
    I18n.t("FIMS.history.exportData.alerts.alreadyExporting.title"),
    I18n.t("FIMS.history.exportData.alerts.alreadyExporting.body"),
    [{ text: I18n.t("global.buttons.ok"), onPress }]
  );

export const useFimsHistoryExport = () => {
  const historyExportState = useIOSelector(fimsHistoryExportStateSelector);
  const dispatch = useIODispatch();
  const [isLoading, setIsLoading] = React.useState(false);

  React.useEffect(() => {
    RemoteValue.fold(
      historyExportState,
      constVoid,
      constVoid,
      value => {
        if (value === "SUCCESS") {
          showFimsExportSuccess();
          setIsLoading(false);
        } else {
          showFimsAlreadyExportingAlert(() => setIsLoading(false));
        }
      },
      () => {
        showFimsExportError();
        setIsLoading(false);
      }
    );
  }, [historyExportState, dispatch]);

  // cleanup
  React.useEffect(
    () => () => {
      dispatch(resetFimsHistoryExportState());
    },
    [dispatch]
  );

  const handleExportOnPress = () =>
    Alert.alert(
      I18n.t("FIMS.history.exportData.alerts.areYouSure"),
      undefined,
      [
        { text: I18n.t("global.buttons.cancel"), style: "cancel" },
        {
          text: I18n.t("global.buttons.confirm"),
          isPreferred: true,
          onPress: () => {
            if (!isLoading) {
              setIsLoading(true);
              dispatch(fimsHistoryExport.request());
            }
          }
        }
      ]
    );

  return {
    handleExportOnPress
  };
};
