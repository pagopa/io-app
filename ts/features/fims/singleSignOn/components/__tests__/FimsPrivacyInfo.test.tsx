import { fireEvent, render } from "@testing-library/react-native";
import type { ComponentProps } from "react";
import I18n from "../../../../../i18n";
import * as UTILS_URL from "../../../../../utils/url";
import { FimsPrivacyInfo } from "../FimsPrivacyInfo";

const privacyUrl = "https://example.com/privacy";
const mockOpenUrl = jest.spyOn(UTILS_URL, "openWebUrl");

describe("FimsPrivacyInfo component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with privacyUrl defined", () => {
    const { toJSON, queryByText, getByText } = renderComponent({ privacyUrl });
    expect(
      queryByText(new RegExp(I18n.t("FIMS.consentsScreen.privacy1")))
    ).toBeTruthy();
    expect(getByText(I18n.t("FIMS.consentsScreen.privacyCta"))).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("render LoadingSkeleton when privacyUrl is undefined", () => {
    const { toJSON, getByTestId } = renderComponent({});
    expect(getByTestId("skeleton")).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("calls openWebUrl with correct URL on press", () => {
    const { getByTestId } = renderComponent({ privacyUrl });
    expect(mockOpenUrl).toHaveBeenCalledTimes(0);

    const link = getByTestId("body-primary-action");
    fireEvent.press(link);

    expect(mockOpenUrl).toHaveBeenCalledWith(privacyUrl);
    expect(mockOpenUrl).toHaveBeenCalledTimes(1);
  });
});

const renderComponent = (props: ComponentProps<typeof FimsPrivacyInfo>) =>
  render(<FimsPrivacyInfo {...props} />);
