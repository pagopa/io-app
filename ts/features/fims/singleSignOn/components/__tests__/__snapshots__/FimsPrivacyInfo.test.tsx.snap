// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FimsPrivacyInfo component render LoadingSkeleton when privacyUrl is undefined 1`] = `
<View
  accessibilityHint="Attendi il caricamento del contenuto"
  accessibilityLabel="Caricamento in corso"
  accessible={true}
  testID="skeleton"
>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
  >
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "column",
          "gap": 8,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#E8EBF1",
            "borderCurve": "continuous",
            "borderRadius": 4,
            "height": 21,
            "opacity": 0.55,
            "width": "98%",
          }
        }
      />
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#E8EBF1",
            "borderCurve": "continuous",
            "borderRadius": 4,
            "height": 21,
            "opacity": 0.55,
            "width": "86%",
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`FimsPrivacyInfo component renders correctly with privacyUrl defined 1`] = `
<Text
  allowFontScaling={true}
  dynamicTypeRamp="body"
  maxFontSizeMultiplier={1.5}
  style={
    [
      {},
      {
        "color": "#555C70",
        "fontFamily": "Titillio",
        "fontSize": 16,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 24,
      },
    ]
  }
>
  Per maggiori informazioni, leggi 
  <View
    accessibilityRole="link"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      accessibilityRole="link"
      allowFontScaling={true}
      dynamicTypeRamp="body"
      maxFontSizeMultiplier={1.5}
      style={
        [
          {
            "textDecorationLine": "underline",
          },
          {
            "color": "#0B3EE3",
            "fontFamily": "Titillio",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "600",
            "lineHeight": 24,
          },
        ]
      }
      testID="body-primary-action"
    >
      lâ€™Informativa sulla Privacy 
    </Text>
  </View>
</Text>
`;
